Lets ENABLE_ENTERPRISE_INTEGRATION be configurable

pomegranited opened the pull request:
Allows an edx-platform instance to disable Enterprise integration via configuration.

**JIRA tickets**: [OSPR-1945](https://openedx.atlassian.net/browse/OSPR-1945)

**Dependencies**: 

Related to https://github.com/edx/configuration/pull/4128

**Sandbox URL**:

* LMS: https://pr16155.sandbox.opencraft.hosting/
* Studio: https://studio-pr16155.sandbox.opencraft.hosting/

Running 425239a.

**Merge deadline**: ASAP

**Testing instructions**:

1. Run the LMS shell, and ensure that the default enterprise integration setting is off by default:
   ```bash
   >>> from openedx.features.enterprise_support.api import enterprise_enabled
   >>> enterprise_enabled()
   False
   ```
1. Update `/edx/app/edxapp/lms.env.json` to add to `"FEATURES"`: 
   ```json
   "ENABLE_ENTERPRISE_INTEGRATION": true,
   ```
1. Restart the LMS shell, and verify that `enterprise_enabled()` now returns `True`.

**Author Notes & Concerns**

We're currently experiencing an error during learner registration via course enrolment, for Ginkgo instances with third-party authentication configured, that sit behind a load balancer and cannot resolve their public Enterprise API URL on https.  The user sees a 500 error.
<details><summary>See traceback</summary>
```
Oct  2 15:50:48 ondemand [service_variant=lms][django.request][env:sandbox] ERROR [ondemand  2723] [base.py:256] - Internal Server Error: /course_modes/choose/course-v1:Cloudera+CN+0/
--
Traceback (most recent call last):
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 132, in get_response
response = wrapped_callback(request, *callback_args, **callback_kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/newrelic/hooks/framework_django.py", line 527, in wrapper
return wrapped(*args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/views/generic/base.py", line 71, in view
return self.dispatch(request, *args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 34, in _wrapper
return bound_func(*args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 30, in bound_func
return func.__get__(self, type(self))(*args2, **kwargs2)
File "/edx/app/edxapp/edx-platform/common/djangoapps/course_modes/views.py", line 56, in dispatch
return super(ChooseModeView, self).dispatch(*args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/newrelic/hooks/framework_django.py", line 914, in wrapper
return wrapped(*args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/views/generic/base.py", line 89, in dispatch
return handler(request, *args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 34, in _wrapper
return bound_func(*args, **kwargs)
File "/edx/app/edxapp/edx-platform/common/djangoapps/third_party_auth/decorators.py", line 87, in inner
return func(request, *args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 30, in bound_func
return func.__get__(self, type(self))(*args2, **kwargs2)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 34, in _wrapper
return bound_func(*args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/contrib/auth/decorators.py", line 22, in _wrapped_view
return view_func(request, *args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 30, in bound_func
return func.__get__(self, type(self))(*args2, **kwargs2)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 34, in _wrapper
return bound_func(*args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 145, in inner
return func(*args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 30, in bound_func
return func.__get__(self, type(self))(*args2, **kwargs2)
File "/edx/app/edxapp/edx-platform/common/djangoapps/course_modes/views.py", line 120, in get
return redirect(self._get_redirect_url_for_audit_enrollment(request, course_id))
File "/edx/app/edxapp/edx-platform/common/djangoapps/course_modes/views.py", line 298, in _get_redirect_url_for_audit_enrollment
enterprise_learner_data = enterprise_api.get_enterprise_learner_data(site=request.site, user=request.user)
File "/edx/app/edxapp/edx-platform/openedx/features/enterprise_support/api.py", line 391, in get_enterprise_learner_data
enterprise_learner_data = EnterpriseApiClient().fetch_enterprise_learner_data(site=site, user=user)
File "/edx/app/edxapp/edx-platform/openedx/features/enterprise_support/api.py", line 205, in fetch_enterprise_learner_data
response = endpoint().get(**querystring)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/slumber/__init__.py", line 155, in get
resp = self._request("GET", params=kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/slumber/__init__.py", line 97, in _request
resp = self._store["session"].request(method, url, data=data, params=params, files=files, headers=headers)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/requests/sessions.py", line 468, in request
resp = self.send(prep, **send_kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/newrelic/api/external_trace.py", line 121, in dynamic_wrapper
return wrapped(*args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/requests/sessions.py", line 576, in send
r = adapter.send(request, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/requests/adapters.py", line 437, in send
raise ConnectionError(e, request=request)
```
</details>

This change allows us to disable Enterprise integration and avoid this issue, but could be addressed in several other ways.

**Reviewers**
- [x] @e-kolpakov 
- [ ] edX reviewer[s] TBD CC @gsong 

**Settings**
```yaml
EDXAPP_FEATURES:
  ENABLE_ENTERPRISE_INTEGRATION: false
```

edx-webhook commented:
Thanks for the pull request, @pomegranited! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=16155&repo=edx%2Fedx-platform"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

edx-webhook commented:
Thanks for the pull request, @pomegranited! I've created [OSPR-1945](https://openedx.atlassian.net/browse/OSPR-1945) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

pomegranited commented:
jenkins run python

feanil commented:
@pomegranited two questions:
- Why default this to true?  It seems like many people in the community would not want it.
- Why not put this setting in the FEATURE_FLAGS dictionary?

pomegranited commented:
Hi @feanil -- thanks for having a look!
> Why default this to true? It seems like many people in the community would not want it.
I am happy to make its default be false, if this is preferred.  I try not to change things when I make configuration changes, just make them configurable.
> Why not put this setting in the FEATURE_FLAGS dictionary?
Good point, I'll raise that option with the Enterprise team.

Thank you!

douglashall commented:
@pomegranited Yes, I agree that we should default the flag to `False`. Also, I agree that this flag should probably be a key in `FEATURE_FLAGS`. Follow these recommendations will result in further refactoring that you may or my not be prepared to take on.

pomegranited commented:
Ach.. thanks @douglashall , that might be true..  I'll check and get back to you.

pomegranited commented:
@douglashall Yep, I've got permission to proceed with making this a feature flag, so I should have something for you soon.  Thanks!

pomegranited commented:
@douglashall I've moved the flag to `settings.FEATURES`, and updated the relevant tests.

Since this will require edx to change a feature flag, what's the best way to get this merged and deployed without disruption?

Thanks for your advice!

pomegranited commented:
Ping @gsong -- this PR is ready for review.

douglashall commented on common/djangoapps/util/tests/test_submit_feedback.py:
```diff
@@ -463,7 +462,8 @@ def test_valid_request_auth_user_with_custom_fields(self, course_id, enrolled, z
     @unpack
-    @override_settings(ZENDESK_CUSTOM_FIELDS=TEST_ZENDESK_CUSTOM_FIELD_CONFIG, ENABLE_ENTERPRISE_INTEGRATION=True)
+    @override_settings(ZENDESK_CUSTOM_FIELDS=TEST_ZENDESK_CUSTOM_FIELD_CONFIG)
+    @mock.patch.dict("django.conf.settings.FEATURES", dict(ENABLE_ENTERPRISE_INTEGRATION=True))
```
Why not use `override_settings` to do this?

pomegranited replied:
@douglashall  I found that if I use `override_settings`, then I end up overriding the whole `FEATURES` block, that that messes with the [`@mock.patch` `FEATURES` that wrap the test class](https://github.com/edx/edx-platform/pull/16155/files#diff-8f395611f456cfd137e4252981eedf31L48).

pomegranited commented:
Thank you @douglashall !  Merging now.  

What do we need to do to ensure that the `ENABLE_ENTERPRISE_INTEGRATION` feature flag gets set before this is deployed to edx.org and edge.edx.org?

FYI @feanil @edx/devops

douglashall commented:
@pomegranited You'll need to add the key in the following locations and set the value to `true`:

https://github.com/edx/edx-internal/blob/e9f1be85455d6853a386fe24705cd2fe8ff310c9/ansible/vars/prod-edx.yml#L159
https://github.com/edx/edx-internal/blob/e9f1be85455d6853a386fe24705cd2fe8ff310c9/ansible/vars/stage-edx.yml#L131

pomegranited commented:
Thanks @douglashall !  Done.

pomegranited commented:
@nedbat Could this be cherry-picked into the ginkgo release branch?  Let me know if you need me to create a PR for that, since there may be conflicts to resolve.

nedbat commented:
@pomegranited I have to admit, this is a bigger change than we would automatically cherry-pick.  But it's mostly test files, and is conceptually simple, so let's go ahead.  A PR would be good, thanks.
